File: ast_proof_system/determinism_analyzer.py
#!/usr/bin/env python3
# SPDX-FileCopyrightText: 2025 Rhett Creighton
# SPDX-License-Identifier: Apache-2.0

"""
Determinism Analyzer for AST GDB Proving System
Tags all operations as deterministic or non-deterministic
Enables stronger proofs and performance optimizations
"""

import subprocess
import re
from dataclasses import dataclass, field
from typing import List, Dict, Set, Optional, Tuple
from enum import Enum
import json

class Determinism(Enum):
    """Determinism classification"""
    DETERMINISTIC = "deterministic"
    NON_DETERMINISTIC = "non_deterministic"
    CONDITIONALLY_DETERMINISTIC = "conditionally_deterministic"
    UNKNOWN = "unknown"

@dataclass
class InputSource:
    """Represents a source of input to the program"""
    name: str
    type: str  # 'user_input', 'time', 'random', 'file', 'network', 'hardware'
    location: str
    determinism: Determinism
    affects_functions: Set[str] = field(default_factory=set)
    taint_paths: List[str] = field(default_factory=list)

@dataclass
class Operation:
    """Enhanced operation with determinism tracking"""
    type: str
    location: str
    function: str
    determinism: Determinism
    input_sources: List[InputSource] = field(default_factory=list)
    proof_strength: str = "weak"  # 'weak', 'strong', 'mathematical'
    optimization_potential: str = "none"  # 'none', 'cache', 'precompute', 'eliminate'

class DeterminismAnalyzer:
    """Analyzes code to identify deterministic vs non-deterministic operations"""
    
    # Non-deterministic sources
    NON_DETERMINISTIC_SOURCES = {
        # User input
        'scanf': 'user_input',
        'getchar': 'user_input',
        'fgets': 'user_input',
        'read': 'user_input',
        'recv': 'network',
        
        # Time-based
        'time': 'time',
        'gettimeofday': 'time',
        'clock': 'time',
        'clock_gettime': 'time',
        
        # Random
        'rand': 'random',
        'random': 'random',
        'drand48': 'random',
        'getrandom': 'random',
        '/dev/random': 'random',
        '/dev/urandom': 'random',
        
        # System state
        'getpid': 'system',
        'getenv': 'environment',
        'getcwd': 'filesystem',
        
        # Hardware
        'rdtsc': 'hardware',  # CPU timestamp counter
        'cpuid': 'hardware',
        
        # Game-specific
        'GetKeyState': 'user_input',
        'GetMousePosition': 'user_input',
        'GetGamepadAxis': 'user_input',
        'NetworkReceive': 'network',
    }
    
    # Deterministic operations (can be proven once)
    DETERMINISTIC_OPERATIONS = {
        'pure_math': ['sin', 'cos', 'sqrt', 'pow', 'fabs'],
        'constants': ['PI', 'E', 'MAX_INT'],
        'compile_time': ['sizeof', 'offsetof', '_Static_assert'],
        'pure_functions': []  # Populated by analysis
    }
    
    def __init__(self, source_file: str):
        self.source_file = source_file
        self.input_sources: List[InputSource] = []
        self.operations: List[Operation] = []
        self.function_determinism: Dict[str, Determinism] = {}
        self.taint_graph: Dict[str, Set[str]] = {}  # var -> sources
        self.current_function = None
        
    def analyze(self) -> Dict:
        """Run complete determinism analysis"""
        print(f"=== Determinism Analysis for {self.source_file} ===")
        
        # Extract AST
        ast_text = self.extract_ast()
        
        # Phase 1: Find all input sources
        self.find_input_sources(ast_text)
        
        # Phase 2: Build taint graph
        self.build_taint_graph(ast_text)
        
        # Phase 3: Classify function determinism
        self.classify_functions(ast_text)
        
        # Phase 4: Tag all operations
        self.tag_operations(ast_text)
        
        # Phase 5: Calculate proof optimization potential
        self.calculate_optimizations()
        
        return self.generate_report()
    
    def extract_ast(self) -> str:
        """Extract AST with full details"""
        cmd = [
            'clang', '-Xclang', '-ast-dump',
            '-Xclang', '-ast-dump-filter=',
            '-fsyntax-only', self.source_file
        ]
        result = subprocess.run(cmd, capture_output=True, text=True)
        return result.stdout
    
    def find_input_sources(self, ast_text: str):
        """Identify all sources of non-determinism"""
        lines = ast_text.split('\n')
        
        for i, line in enumerate(lines):
            # Track current function
            if 'FunctionDecl' in line:
                match = re.search(r"'([^']+)'", line)
                if match:
                    self.current_function = match.group(1)
            
            # Check for non-deterministic calls
            if 'CallExpr' in line:
                for func_name, input_type in self.NON_DETERMINISTIC_SOURCES.items():
                    if func_name in line:
                        location = self.extract_location(line)
                        source = InputSource(
                            name=func_name,
                            type=input_type,
                            location=location,
                            determinism=Determinism.NON_DETERMINISTIC
                        )
                        source.affects_functions.add(self.current_function or 'global')
                        self.input_sources.append(source)
                        
            # Check for volatile variables (hardware access)
            elif 'volatile' in line:
                location = self.extract_location(line)
                source = InputSource(
                    name='volatile_access',
                    type='hardware',
                    location=location,
                    determinism=Determinism.NON_DETERMINISTIC
                )
                source.affects_functions.add(self.current_function or 'global')
                self.input_sources.append(source)
    
    def build_taint_graph(self, ast_text: str):
        """Track how non-determinism propagates through the program"""
        # Simplified taint analysis
        # In production, would use proper dataflow analysis
        
        tainted_vars = set()
        
        # Mark variables assigned from non-deterministic sources
        for source in self.input_sources:
            # Find assignments from this source
            # This is simplified - real implementation would parse AST properly
            if source.affects_functions:
                for func in source.affects_functions:
                    self.taint_graph[func] = self.taint_graph.get(func, set())
                    self.taint_graph[func].add(source.name)
    
    def classify_functions(self, ast_text: str):
        """Classify each function's determinism"""
        lines = ast_text.split('\n')
        current_func = None
        func_calls = {}  # function -> called functions
        
        for line in lines:
            if 'FunctionDecl' in line:
                match = re.search(r"'([^']+)'", line)
                if match:
                    current_func = match.group(1)
                    self.function_determinism[current_func] = Determinism.DETERMINISTIC
                    func_calls[current_func] = set()
            
            elif 'CallExpr' in line and current_func:
                # Track function calls
                match = re.search(r"'([^']+)'", line)
                if match:
                    called = match.group(1)
                    func_calls[current_func].add(called)
        
        # Propagate non-determinism through call graph
        changed = True
        while changed:
            changed = False
            for func, calls in func_calls.items():
                if self.function_determinism.get(func) == Determinism.DETERMINISTIC:
                    # Check if any called function is non-deterministic
                    for called in calls:
                        if called in self.NON_DETERMINISTIC_SOURCES:
                            self.function_determinism[func] = Determinism.NON_DETERMINISTIC
                            changed = True
                            break
                        elif self.function_determinism.get(called) == Determinism.NON_DETERMINISTIC:
                            self.function_determinism[func] = Determinism.NON_DETERMINISTIC
                            changed = True
                            break
        
        # Check for tainted functions
        for func, sources in self.taint_graph.items():
            if sources:
                self.function_determinism[func] = Determinism.NON_DETERMINISTIC
    
    def tag_operations(self, ast_text: str):
        """Tag each dangerous operation with determinism info"""
        # Import enhanced analyzer
        from enhanced_ast_analyzer import EnhancedASTAnalyzer
        
        analyzer = EnhancedASTAnalyzer(self.source_file)
        results = analyzer.analyze_complete()
        
        # Tag each dangerous operation
        for dangerous_op in results['dangerous_operations']:
            op = Operation(
                type=dangerous_op.type,
                location=dangerous_op.location,
                function=dangerous_op.function,
                determinism=self.function_determinism.get(
                    dangerous_op.function, 
                    Determinism.UNKNOWN
                )
            )
            
            # Determine proof strength based on determinism
            if op.determinism == Determinism.DETERMINISTIC:
                op.proof_strength = "mathematical"
                op.optimization_potential = "cache"
            elif op.determinism == Determinism.NON_DETERMINISTIC:
                op.proof_strength = "weak"
                op.optimization_potential = "none"
            
            # Special cases
            if op.type == 'division' and op.determinism == Determinism.DETERMINISTIC:
                op.optimization_potential = "precompute"
            elif op.type == 'array_bounds' and op.determinism == Determinism.DETERMINISTIC:
                op.optimization_potential = "eliminate"  # Can be removed after first proof
            
            self.operations.append(op)
    
    def calculate_optimizations(self):
        """Calculate optimization potential based on determinism"""
        self.optimization_stats = {
            'total_operations': len(self.operations),
            'deterministic_ops': 0,
            'non_deterministic_ops': 0,
            'cacheable_proofs': 0,
            'eliminable_checks': 0,
            'precomputable_values': 0,
            'estimated_speedup': 0.0
        }
        
        for op in self.operations:
            if op.determinism == Determinism.DETERMINISTIC:
                self.optimization_stats['deterministic_ops'] += 1
                
                if op.optimization_potential == "cache":
                    self.optimization_stats['cacheable_proofs'] += 1
                elif op.optimization_potential == "eliminate":
                    self.optimization_stats['eliminable_checks'] += 1
                elif op.optimization_potential == "precompute":
                    self.optimization_stats['precomputable_values'] += 1
            else:
                self.optimization_stats['non_deterministic_ops'] += 1
        
        # Estimate speedup
        if self.optimization_stats['total_operations'] > 0:
            deterministic_ratio = (self.optimization_stats['deterministic_ops'] / 
                                 self.optimization_stats['total_operations'])
            # Rough estimate: deterministic ops can be 10x faster with caching
            self.optimization_stats['estimated_speedup'] = 1 + (deterministic_ratio * 9)
    
    def extract_location(self, line: str) -> str:
        """Extract file:line location from AST line"""
        match = re.search(r'<([^>]+)>', line)
        if match:
            return match.group(1).split(',')[0]
        return "unknown"
    
    def generate_report(self) -> Dict:
        """Generate comprehensive determinism report"""
        report = {
            'source_file': self.source_file,
            'input_sources': [
                {
                    'name': src.name,
                    'type': src.type,
                    'location': src.location,
                    'affects': list(src.affects_functions)
                }
                for src in self.input_sources
            ],
            'function_determinism': {
                func: det.value 
                for func, det in self.function_determinism.items()
            },
            'operations': [
                {
                    'type': op.type,
                    'location': op.location,
                    'function': op.function,
                    'determinism': op.determinism.value,
                    'proof_strength': op.proof_strength,
                    'optimization': op.optimization_potential
                }
                for op in self.operations
            ],
            'optimization_potential': self.optimization_stats,
            'recommendations': self.generate_recommendations()
        }
        
        return report
    
    def generate_recommendations(self) -> List[str]:
        """Generate optimization recommendations"""
        recommendations = []
        
        if self.optimization_stats['deterministic_ops'] > 0:
            recommendations.append(
                f"Enable proof caching for {self.optimization_stats['deterministic_ops']} "
                f"deterministic operations (estimated {self.optimization_stats['estimated_speedup']:.1f}x speedup)"
            )
        
        if self.optimization_stats['eliminable_checks'] > 0:
            recommendations.append(
                f"Consider compile-time elimination of {self.optimization_stats['eliminable_checks']} "
                "redundant safety checks in deterministic code paths"
            )
        
        if self.optimization_stats['precomputable_values'] > 0:
            recommendations.append(
                f"Precompute {self.optimization_stats['precomputable_values']} "
                "constant expressions at compile time"
            )
        
        # Suggest isolation of non-deterministic code
        if self.optimization_stats['non_deterministic_ops'] > 0:
            recommendations.append(
                "Isolate non-deterministic operations into separate functions "
                "to maximize deterministic code regions"
            )
        
        return recommendations

def generate_optimized_prover(report: Dict) -> str:
    """Generate optimized GDB prover based on determinism analysis"""
    prover = [
        "# Optimized GDB Prover with Determinism Awareness",
        "# Generated from determinism analysis",
        "",
        "# Cache for deterministic proofs",
        "set $proof_cache = {}",
        "",
        "define verify_with_cache",
        "  set $op_location = $arg0",
        "  set $op_determinism = $arg1",
        "  ",
        "  # Check cache for deterministic operations",
        '  if $op_determinism == "deterministic"',
        "    # Look up in cache",
        "    set $cached = 0  # Would check actual cache",
        "    if $cached",
        '      printf "CACHED PROOF: %s is safe\\n", $op_location',
        "      return",
        "    end",
        "  end",
        "  ",
        "  # Perform actual verification",
        '  printf "VERIFYING: %s\\n", $op_location',
        "  # ... verification logic ...",
        "  ",
        "  # Cache result if deterministic",
        '  if $op_determinism == "deterministic"',
        "    # Store in cache",
        '    printf "CACHING proof for %s\\n", $op_location',
        "  end",
        "end",
        ""
    ]
    
    # Add specific verification for each operation
    for op in report['operations']:
        if op['determinism'] == 'deterministic':
            prover.append(f'# {op["location"]}: {op["type"]} - DETERMINISTIC (cached)')
            prover.append(f'verify_with_cache "{op["location"]}" "deterministic"')
        else:
            prover.append(f'# {op["location"]}: {op["type"]} - NON-DETERMINISTIC (always verify)')
            prover.append(f'verify_with_cache "{op["location"]}" "non_deterministic"')
        prover.append("")
    
    return '\n'.join(prover)

def main():
    import sys
    
    if len(sys.argv) < 2:
        print("Usage: determinism_analyzer.py <source_file.c>")
        sys.exit(1)
    
    analyzer = DeterminismAnalyzer(sys.argv[1])
    report = analyzer.analyze()
    
    # Print summary
    print("\n=== Determinism Analysis Summary ===")
    print(f"Input sources found: {len(report['input_sources'])}")
    for src in report['input_sources']:
        print(f"  - {src['name']} ({src['type']}) at {src['location']}")
    
    print(f"\nFunction determinism:")
    det_funcs = sum(1 for d in report['function_determinism'].values() 
                    if d == 'deterministic')
    non_det_funcs = sum(1 for d in report['function_determinism'].values() 
                        if d == 'non_deterministic')
    print(f"  Deterministic: {det_funcs}")
    print(f"  Non-deterministic: {non_det_funcs}")
    
    print(f"\nOptimization potential:")
    opt = report['optimization_potential']
    print(f"  Total operations: {opt['total_operations']}")
    print(f"  Deterministic: {opt['deterministic_ops']} ({opt['deterministic_ops']/opt['total_operations']*100:.1f}%)")
    print(f"  Cacheable proofs: {opt['cacheable_proofs']}")
    print(f"  Eliminable checks: {opt['eliminable_checks']}")
    print(f"  Estimated speedup: {opt['estimated_speedup']:.1f}x")
    
    print(f"\nRecommendations:")
    for rec in report['recommendations']:
        print(f"  • {rec}")
    
    # Save detailed report
    with open('determinism_report.json', 'w') as f:
        json.dump(report, f, indent=2)
    print(f"\nDetailed report saved to: determinism_report.json")
    
    # Generate optimized prover
    prover_code = generate_optimized_prover(report)
    with open('optimized_prover.gdb', 'w') as f:
        f.write(prover_code)
    print(f"Optimized GDB prover saved to: optimized_prover.gdb")

if __name__ == '__main__':
    main()

File: ast_proof_system/src/determinism_analyzer.c
/* SPDX-FileCopyrightText: 2025 Rhett Creighton
 * SPDX-License-Identifier: Apache-2.0
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdbool.h>

#include "determinism_analyzer.h"

/* Non-deterministic function database */
typedef struct {
    const char* name;
    input_source_t source;
    const char* description;
} nondet_function_t;

static const nondet_function_t nondet_functions[] = {
    /* User input */
    {"scanf", INPUT_USER, "User input"},
    {"fscanf", INPUT_USER, "File input"},
    {"getchar", INPUT_USER, "Character input"},
    {"gets", INPUT_USER, "String input"},
    {"fgets", INPUT_USER, "Line input"},
    {"read", INPUT_USER, "Low-level input"},
    
    /* Time */
    {"time", INPUT_TIME, "Current time"},
    {"clock", INPUT_TIME, "Processor time"},
    {"gettimeofday", INPUT_TIME, "High-res time"},
    
    /* Random */
    {"rand", INPUT_RANDOM, "Random number"},
    {"random", INPUT_RANDOM, "Random number"},
    {"drand48", INPUT_RANDOM, "Random double"},
    
    /* System */
    {"getenv", INPUT_SYSTEM, "Environment variable"},
    {"getpid", INPUT_SYSTEM, "Process ID"},
    {"getcwd", INPUT_SYSTEM, "Current directory"},
    
    /* Network */
    {"recv", INPUT_NETWORK, "Network receive"},
    {"recvfrom", INPUT_NETWORK, "UDP receive"},
    {"accept", INPUT_NETWORK, "Accept connection"},
    
    {NULL, INPUT_DETERMINISTIC, NULL}
};

/* Initialize determinism analyzer */
void init_determinism_analyzer(determinism_analyzer_t* analyzer) {
    memset(analyzer, 0, sizeof(*analyzer));
}

/* Check if function is non-deterministic */
static input_source_t check_function_determinism(const char* func_name) {
    for (int i = 0; nondet_functions[i].name != NULL; i++) {
        if (strcmp(func_name, nondet_functions[i].name) == 0) {
            return nondet_functions[i].source;
        }
    }
    return INPUT_DETERMINISTIC;
}

/* Analyze function for determinism */
int analyze_function_determinism(determinism_analyzer_t* analyzer,
                                const char* func_name,
                                const ast_node_t* nodes,
                                int node_count) {
    if (analyzer->function_count >= MAX_FUNCTIONS) {
        return -1;
    }
    
    function_info_t* func = &analyzer->functions[analyzer->function_count];
    strncpy(func->name, func_name, sizeof(func->name) - 1);
    func->determinism = DETERMINISM_DETERMINISTIC;
    func->input_sources = INPUT_DETERMINISTIC;
    
    /* Scan nodes for non-deterministic calls */
    for (int i = 0; i < node_count; i++) {
        if (nodes[i].type == 3) { /* AST_CALL_EXPR */
            /* Extract called function name */
            char called_func[128] = {0};
            const char* start = strstr(nodes[i].content, "'");
            if (start) {
                const char* end = strchr(start + 1, '\'');
                if (end) {
                    size_t len = end - start - 1;
                    if (len < sizeof(called_func) - 1) {
                        strncpy(called_func, start + 1, len);
                    }
                }
            }
            
            /* Check if non-deterministic */
            input_source_t source = check_function_determinism(called_func);
            if (source != INPUT_DETERMINISTIC) {
                func->determinism = DETERMINISM_NON_DETERMINISTIC;
                func->input_sources |= source;
                func->nondet_call_count++;
            }
        }
    }
    
    analyzer->function_count++;
    
    /* Update statistics */
    if (func->determinism == DETERMINISM_DETERMINISTIC) {
        analyzer->deterministic_count++;
    } else {
        analyzer->non_deterministic_count++;
    }
    
    return 0;
}

/* Propagate determinism through call graph */
void propagate_determinism(determinism_analyzer_t* analyzer) {
    bool changed = true;
    
    while (changed) {
        changed = false;
        
        for (int i = 0; i < analyzer->function_count; i++) {
            function_info_t* caller = &analyzer->functions[i];
            
            if (caller->determinism == DETERMINISM_DETERMINISTIC) {
                /* Check if any called function is non-deterministic */
                for (int j = 0; j < caller->call_count; j++) {
                    const char* callee_name = caller->calls[j];
                    
                    /* Find callee */
                    for (int k = 0; k < analyzer->function_count; k++) {
                        if (strcmp(analyzer->functions[k].name, callee_name) == 0) {
                            if (analyzer->functions[k].determinism == DETERMINISM_NON_DETERMINISTIC) {
                                /* Propagate non-determinism */
                                caller->determinism = DETERMINISM_NON_DETERMINISTIC;
                                caller->input_sources |= analyzer->functions[k].input_sources;
                                analyzer->deterministic_count--;
                                analyzer->non_deterministic_count++;
                                changed = true;
                                break;
                            }
                        }
                    }
                    
                    if (changed) break;
                }
            }
        }
    }
}

/* Analyze operations for optimization potential */
void analyze_operation_optimization(determinism_analyzer_t* analyzer,
                                  const dangerous_op_t* op,
                                  operation_optimization_t* opt) {
    /* Find function containing operation */
    function_info_t* func = NULL;
    for (int i = 0; i < analyzer->function_count; i++) {
        if (strcmp(analyzer->functions[i].name, op->function) == 0) {
            func = &analyzer->functions[i];
            break;
        }
    }
    
    if (!func) {
        opt->strategy = OPT_NONE;
        return;
    }
    
    /* Determine optimization based on determinism */
    if (func->determinism == DETERMINISM_DETERMINISTIC) {
        if (op->is_safe) {
            opt->strategy = OPT_ELIMINATE;  /* Can remove after first proof */
            opt->speedup_factor = 10.0;
        } else {
            opt->strategy = OPT_CACHE;  /* Cache the proof */
            opt->speedup_factor = 5.0;
        }
        opt->can_cache = true;
        opt->can_eliminate = op->is_safe;
    } else {
        /* Non-deterministic operations */
        if (op->is_safe && (func->input_sources & INPUT_USER) == 0) {
            /* Bounded non-determinism (e.g., time-based) */
            opt->strategy = OPT_CACHE;
            opt->speedup_factor = 2.0;
            opt->can_cache = true;
        } else {
            /* Unbounded non-determinism */
            opt->strategy = OPT_NONE;
            opt->speedup_factor = 1.0;
        }
        opt->can_eliminate = false;
    }
}

/* Calculate overall optimization potential */
void calculate_optimization_stats(determinism_analyzer_t* analyzer) {
    optimization_stats_t* stats = &analyzer->opt_stats;
    
    stats->total_operations = analyzer->total_operations;
    
    double total_original_cost = 0.0;
    double total_optimized_cost = 0.0;
    
    for (int i = 0; i < analyzer->operation_count; i++) {
        const operation_optimization_t* opt = &analyzer->operations[i];
        
        /* Assume base cost of 1.0 per operation */
        total_original_cost += 1.0;
        
        switch (opt->strategy) {
        case OPT_ELIMINATE:
            stats->eliminable_operations++;
            total_optimized_cost += 0.0;  /* No cost after elimination */
            break;
            
        case OPT_CACHE:
            stats->cacheable_operations++;
            total_optimized_cost += 0.2;  /* Cache lookup cost */
            break;
            
        case OPT_PRECOMPUTE:
            stats->precomputable_operations++;
            total_optimized_cost += 0.1;  /* Precomputed value access */
            break;
            
        default:
            total_optimized_cost += 1.0;  /* Full cost */
            break;
        }
    }
    
    if (total_original_cost > 0) {
        stats->estimated_speedup = total_original_cost / total_optimized_cost;
    } else {
        stats->estimated_speedup = 1.0;
    }
}

/* Print determinism analysis */
void print_determinism_analysis(const determinism_analyzer_t* analyzer) {
    printf("=== Determinism Analysis ===\n");
    printf("Functions analyzed: %d\n", analyzer->function_count);
    printf("  Deterministic: %d\n", analyzer->deterministic_count);
    printf("  Non-deterministic: %d\n", analyzer->non_deterministic_count);
    printf("\n");
    
    /* Show non-deterministic functions */
    if (analyzer->non_deterministic_count > 0) {
        printf("Non-deterministic functions:\n");
        for (int i = 0; i < analyzer->function_count; i++) {
            const function_info_t* func = &analyzer->functions[i];
            if (func->determinism == DETERMINISM_NON_DETERMINISTIC) {
                printf("  %s: ", func->name);
                
                /* Show input sources */
                if (func->input_sources & INPUT_USER) printf("USER ");
                if (func->input_sources & INPUT_TIME) printf("TIME ");
                if (func->input_sources & INPUT_RANDOM) printf("RANDOM ");
                if (func->input_sources & INPUT_SYSTEM) printf("SYSTEM ");
                if (func->input_sources & INPUT_NETWORK) printf("NETWORK ");
                
                printf("\n");
            }
        }
    }
    
    /* Show optimization potential */
    const optimization_stats_t* stats = &analyzer->opt_stats;
    printf("\nOptimization potential:\n");
    printf("  Eliminable operations: %d\n", stats->eliminable_operations);
    printf("  Cacheable operations: %d\n", stats->cacheable_operations);
    printf("  Estimated speedup: %.1fx\n", stats->estimated_speedup);
}

/* Generate optimized verification code */
int generate_optimized_verification(const determinism_analyzer_t* analyzer,
                                   const char* output_file) {
    FILE* fp = fopen(output_file, "w");
    if (!fp) return -1;
    
    fprintf(fp, "/* SPDX-FileCopyrightText: 2025 Rhett Creighton\n");
    fprintf(fp, " * SPDX-License-Identifier: Apache-2.0\n");
    fprintf(fp, " */\n\n");
    
    fprintf(fp, "/* Optimized verification based on determinism analysis */\n\n");
    
    fprintf(fp, "#include <stdbool.h>\n");
    fprintf(fp, "#include <stdint.h>\n\n");
    
    /* Generate cache structure */
    fprintf(fp, "/* Proof cache for deterministic operations */\n");
    fprintf(fp, "typedef struct {\n");
    fprintf(fp, "    uint32_t operation_id;\n");
    fprintf(fp, "    bool verified;\n");
    fprintf(fp, "} proof_cache_entry_t;\n\n");
    
    fprintf(fp, "#define PROOF_CACHE_SIZE 1024\n");
    fprintf(fp, "static proof_cache_entry_t proof_cache[PROOF_CACHE_SIZE];\n");
    fprintf(fp, "static int cache_count = 0;\n\n");
    
    /* Generate verification functions */
    fprintf(fp, "/* Check if operation needs verification */\n");
    fprintf(fp, "static bool needs_verification(uint32_t op_id, bool is_deterministic) {\n");
    fprintf(fp, "    if (!is_deterministic) return true;\n");
    fprintf(fp, "    \n");
    fprintf(fp, "    /* Check cache */\n");
    fprintf(fp, "    for (int i = 0; i < cache_count; i++) {\n");
    fprintf(fp, "        if (proof_cache[i].operation_id == op_id) {\n");
    fprintf(fp, "            return !proof_cache[i].verified;\n");
    fprintf(fp, "        }\n");
    fprintf(fp, "    }\n");
    fprintf(fp, "    return true;\n");
    fprintf(fp, "}\n\n");
    
    /* Generate specific verification macros */
    fprintf(fp, "/* Optimized verification macros */\n");
    
    int eliminable = 0;
    for (int i = 0; i < analyzer->operation_count; i++) {
        const operation_optimization_t* opt = &analyzer->operations[i];
        if (opt->strategy == OPT_ELIMINATE) {
            fprintf(fp, "#define VERIFY_OP_%d() /* Eliminated - proven safe */\n", i);
            eliminable++;
        } else if (opt->strategy == OPT_CACHE) {
            fprintf(fp, "#define VERIFY_OP_%d() do { \\\n", i);
            fprintf(fp, "    if (needs_verification(%d, true)) { \\\n", i);
            fprintf(fp, "        /* Perform verification */ \\\n");
            fprintf(fp, "        proof_cache[cache_count++] = (proof_cache_entry_t){%d, true}; \\\n", i);
            fprintf(fp, "    } \\\n");
            fprintf(fp, "} while(0)\n");
        } else {
            fprintf(fp, "#define VERIFY_OP_%d() /* Runtime verification required */\n", i);
        }
    }
    
    fprintf(fp, "\n/* Total eliminated checks: %d */\n", eliminable);
    
    fclose(fp);
    return 0;
}